hi-ml-histopathology/src/histopathology/utils/report_utils.py

```diff
@@ -0,0 +1,339 @@
+    """Download any file from an AML run if it doesn't exist locally.
+
+    :param run: AML Run object.
+    :param remote_dir: Remote directory from where the file is downloaded.
+    :param download_dir: Local directory where to save the downloaded file.
+    :param filename: Name of the file to be downloaded (e.g. `"test_output.csv"`).
+    :return: Local path to the downloaded file.
+    """
+    local_path = download_dir / filename
+    remote_path = remote_dir / filename
+    if local_path.exists():
+        print("File already exists at", local_path)
+    else:
+        local_path.parent.mkdir(exist_ok=True, parents=True)
+        run.download_file(str(remote_path), str(local_path), _validate_checksum=True)
+        assert local_path.exists()
+        print("File is downloaded at", local_path)
+    return local_path
+
+
+def collect_crossval_outputs(parent_run_id: str, download_dir: Path, aml_workspace: Workspace,
+                             crossval_arg_name: str = "cross_validation_split_index",
+                             output_filename: str = "test_output.csv") -> Dict[int, pd.DataFrame]:
+    """Fetch output CSV files from cross-validation runs as dataframes.
+
+    Will only download the CSV files if they do not already exist locally.
+
+    :param parent_run_id: Azure ML run ID for the parent Hyperdrive run.
+    :param download_dir: Base directory where to download the CSV files. A new sub-directory will
+        be created for each child run (e.g. `<download_dir>/<crossval index>/*.csv`).
+    :param aml_workspace: Azure ML workspace in which the runs were executed.
+    :param crossval_arg_name: Name of the Hyperdrive argument used for indexing the child runs.
+    :param output_filename: Filename of the output CSVs to download.
+    :return: A dictionary of dataframes with the sorted cross-validation indices as keys.
+    """
+    parent_run = get_aml_run_from_run_id(parent_run_id, aml_workspace)
+
+    all_outputs_dfs = {}
+    for child_run in parent_run.get_children():
+        child_run_index = get_tags_from_hyperdrive_run(child_run, crossval_arg_name)
+        if child_run_index is None:
+            raise ValueError(f"Child run expected to have the tag '{crossval_arg_name}'")
+        child_dir = download_dir / str(child_run_index)
+        try:
+            child_csv = download_from_run_if_necessary(child_run,
+                                                       remote_dir=Path("outputs"),
+                                                       download_dir=child_dir,
+                                                       filename=output_filename)
+            all_outputs_dfs[child_run_index] = pd.read_csv(child_csv)
+        except Exception as e:
+            print(f"Failed to download {output_filename} for run {child_run.id}: {e}")
+    return dict(sorted(all_outputs_dfs.items()))
+
+
+def collect_crossval_metrics(parent_run_id: str, download_dir: Path, aml_workspace: Workspace,
```

Comment:
This also looks useful, could live in health_ml.utils to keep it together with `aggregate_hyperdrive_metrics` ? (Would need to replace references to crossval with the more generic 'hyperdrive')

Reply:
This is just a shallow wrapper around `aggregate_hyperdrive_metrics()` to avoid repeatedly re-downloading metrics. Now added `overwrite` flag, changed to JSON serialisation, and wrote a missing test for this function.